contracts/misc/P2PPairStaking.sol

```diff
@@ -0,0 +1,586 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.10;
+
+import "../dependencies/openzeppelin/upgradeability/Initializable.sol";
+import "../dependencies/openzeppelin/upgradeability/OwnableUpgradeable.sol";
+import "../dependencies/yoga-labs/ApeCoinStaking.sol";
+import "../interfaces/IP2PPairStaking.sol";
+import "../dependencies/openzeppelin/contracts/SafeCast.sol";
+import "../interfaces/IAutoCompoundApe.sol";
+import "../interfaces/ICApe.sol";
+import {IERC721} from "../dependencies/openzeppelin/contracts/IERC721.sol";
+import {IERC20, SafeERC20} from "../dependencies/openzeppelin/contracts/SafeERC20.sol";
+import {PercentageMath} from "../protocol/libraries/math/PercentageMath.sol";
+import {SignatureChecker} from "../dependencies/looksrare/contracts/libraries/SignatureChecker.sol";
+
+contract P2PPairStaking is Initializable, OwnableUpgradeable, IP2PPairStaking {
+    using SafeERC20 for IERC20;
+    using PercentageMath for uint256;
+    using SafeCast for uint256;
+
+    bytes32 public constant LISTING_ORDER_HASH =
+        keccak256(
+            "ListingOrder(uint8 stakingType,address offerer,address token,uint256 tokenId,uint256 share,uint256 startTime,uint256 endTime)"
+        );
+    bytes32 public constant MATCHED_ORDER_HASH =
+        keccak256(
+            "MatchedOrder(uint8 stakingType,address apeToken,address apeOfferer,uint32 apeTokenId,uint32 apeShare,address bakcOfferer,uint32 bakcTokenId,uint32 bakcShare,address apeCoinOfferer,uint32 apeCoinShare,uint256 apePrincipleAmount)"
+        );
+    bytes32 internal constant EIP712_DOMAIN =
+        keccak256(
+            "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+        );
+
+    address public immutable bayc;
+    address public immutable mayc;
+    address public immutable bakc;
+    address public immutable apeCoin;
+    address public immutable cApe;
+    ApeCoinStaking public immutable apeCoinStaking;
+
+    bytes32 public DOMAIN_SEPARATOR;
+    mapping(bytes32 => bool) public isListingOrderCanceled;
+    mapping(bytes32 => MatchedOrder) public matchedOrders;
+    mapping(address => uint256) public cApeShareBalance;
+    address public matchingOperator;
+
+    constructor(
+        address _bayc,
+        address _mayc,
+        address _bakc,
+        address _apeCoin,
+        address _cApe,
+        address _apeCoinStaking
+    ) {
+        bayc = _bayc;
+        mayc = _mayc;
+        bakc = _bakc;
+        apeCoin = _apeCoin;
+        cApe = _cApe;
+        apeCoinStaking = ApeCoinStaking(_apeCoinStaking);
```

Comment:
shall we get bayc, mayc, bakc, apeCoin from apeCoinStaking? which may makes the deployment script easier

Reply:
I think it's ok. and i'll cost some gas to fetch this value. more over it has strong requirement for us to use the same bayc, mayc, bakc and apeCoin with apeCoinStaking.